### ISSUE TITLE
Bind mounts of stale FUSE file systems remain despite auto_unmount

### ISSUE THREAD

**Issue by GigabyteProductions** [libfuse/libfuse#889]
I know auto_unmount spawns a fusermount process to unmount the stale mount point after the parent dies. Perhaps fusermount should use the same facility as used by `findmnt` to discover other fuse mounts with the same device ID as the original mount?

Example of such info even after `killall -9 sshfs` left all of these mounts in "Transport endpoint is not connected" state:
```
$ findmnt -Al -o TARGET,FSTYPE,MAJ:MIN
...
/var/tmp/sshfs/a                                                         fuse.sshfs    0:64 
/var/tmp/sshfs/a.bind                                                    fuse.sshfs    0:64 
/var/tmp/sshfs/b                                                         fuse.sshfs    0:65 
```

If mounted with `auto_unmount`, it is just `a` and `b` that get unmounted after a `killall -9 sshfs`, while `a.bind` remains, but `a.bind` gets the same device ID as `a` due to the bind mount.

**Comment by GigabyteProductions** [comment]
I believe this information comes from `/proc/self/mountinfo`

**Comment by bsbernd** [comment]
You might be able to improve things, but this would be still racy - as bind mount might be created while the daemon already in while libfuse tries to unmount.

**Comment by bsbernd** [comment]
If you want to write a patch for it, please go for it.

**Comment by GigabyteProductions** [comment]
I'm actually surprised to see that I can still bind mount specifically from the root of a stale mount; I thought this would fail!

Am I correct in believing that, once there are no more mounted file systems with the original device ID, another bind mount to the stale FS is impossible to create?

Furthermore, is there any existing way to get the kernel itself to immediately tear down the mount point in response to "Transport endpoint is not connected", and redirect the FS call to the FS below? If so, it may help solve the auto_unmount+bind problem, and also eliminate programs getting "Transport endpoint is not connected" during the brief period of time between the FUSE process failing and fusermount actually unmounting the stale moint point(s).

**Comment by bsbernd** [comment]
Problem is that I don't think you can trust the device ID either - I these numbers get reused. That is for sure not a 64-bit ID... Try it out yourself and then check `/sys/fs/fuse/connections`. Then kill your daemon, check what has disappeared, then recreate your daemon - likely the same number reappears. And reappearing number means you cannot trust it, because it might be a new mount. 

I also don't think there is anything in kernel that would help you. The auto-umount thing is rather libfuse specific. Kernel side just sees that the reference to the /dev/fuse file descriptor disappeared and then runs `fuse_dev_release` in fuse-kernel. But that has nothing to do with the mount (which is vfs), but it purely fuse specific. 

In summary, in my opinion you ask for an impossible fix. What you would need would be a unique mount ID, with the same value for bind mounts. Kernel size got unique mount IDs last year, but then every mount point has its down ID - would not help you either.

**Comment by GigabyteProductions** [comment]
Hmmm... interesting... although I'm observing that the inode number changes for the directories in `/sys/fs/fuse/connections` after they are reused:

```
kmarek@kyle-ppc64le ~
$ awk '($5=="/var/tmp/sshfs/a"){print $3;}' /proc/self/mountinfo

kmarek@kyle-ppc64le ~
$ sshfs -o allow_other localhost:/ /var/tmp/sshfs/a

kmarek@kyle-ppc64le ~
$ awk '($5=="/var/tmp/sshfs/a"){print $3;}' /proc/self/mountinfo
0:64

kmarek@kyle-ppc64le ~
$ stat -c %i /sys/fs/fuse/connections/64
3346484

kmarek@kyle-ppc64le ~
$ stat -c %i /sys/fs/fuse/connections/64
3346484

kmarek@kyle-ppc64le ~
$ killall -9 sshfs

kmarek@kyle-ppc64le ~
$ stat -c %i /sys/fs/fuse/connections/64
3346484

kmarek@kyle-ppc64le ~
$ fusermount -u /var/tmp/sshfs/a

kmarek@kyle-ppc64le ~
$ awk '($5=="/var/tmp/sshfs/a"){print $3;}' /proc/self/mountinfo

kmarek@kyle-ppc64le ~
$ sshfs -o allow_other localhost:/ /var/tmp/sshfs/a

kmarek@kyle-ppc64le ~
$ awk '($5=="/var/tmp/sshfs/a"){print $3;}' /proc/self/mountinfo
0:64

kmarek@kyle-ppc64le ~
$ stat -c %i /sys/fs/fuse/connections/64
3344566

kmarek@kyle-ppc64le ~
$ stat -c %i /sys/fs/fuse/connections/64
3344566
```

I'll have to do more digging to see where the inode number under /sys comes from to make sure I'm not just observing effects of something like file system caching.

**Comment by bsbernd** [comment]
I don't think you want to rely on sysfs inode numbers. That might work today and break with kernfs updates. Nobody would expect that you use sysfs inode numbers for fuse autoumount logic.

**Comment by bsbernd** [comment]
I think best would be to ignore racing bind mounts and just to umount everything that the first run finds. And to document that limitation.
